lib/xapi-stdext-std/xstringext_test.ml

```diff
@@ -0,0 +1,187 @@
+      (0, [('.', "...", ["..."]); ('.', "foo.bar.baz", ["foo.bar.baz"])])
+    ; (1, [('.', "...", ["..."]); ('.', "foo.bar.baz", ["foo.bar.baz"])])
+    ; (2, [('.', "...", [""; ".."]); ('.', "foo.bar.baz", ["foo"; "bar.baz"])])
+    ; ( 3
+      , [
+          ('.', "...", [""; ""; "."])
+        ; ('.', "foo.bar.baz", ["foo"; "bar"; "baz"])
+        ] )
+    ; (4, [('.', "...", [""; ""; ""; ""])])
+    ]
+  in
+  let tests_limit =
+    List.map (fun (limit, spec) -> List.map (test ~limit) spec) specs_limit
+    |> List.concat
+  in
+  ("split", List.concat [tests_no_limit; tests_limit])
+
+let test_split_f =
+  let specs =
+    [
+      (XString.isspace, "foo bar", ["foo"; "bar"]) ;
+      (XString.isspace, "foo  bar", ["foo"; "bar"]) ;
+      (XString.isspace, "foo \n\t\r bar", ["foo"; "bar"]) ;
+      (XString.isspace, " foo bar ", ["foo"; "bar"]) ;
+      (XString.isspace, "", []);
+      (XString.isspace, " ", [])
+    ]
+  in
+  let test (splitter, splitted, expected) =
+    let name = Printf.sprintf {|"%s"|} (String.escaped splitted) in
+    test_list (XString.split_f splitter) (name, splitted, expected)
+  in
+  let tests = List.map test specs in
+  ("split_f", tests)
+
+let test_has_substr =
+  let spec =
+    [
+      ("", "", true)
+    ; ("", "foo bar", true)
+    ; ("f", "foof", true)
+    ; ("foofo", "foof", false)
+    ; ("foof", "foof", true)
+    ; ("f", "foof", true)
+    ; ("fo", "foof", true)
+    ; ("of", "foof", true)
+    ; ("ff", "foof", false)
+    ]
+  in
+  let test (contained, container, expected) =
+    let name = Printf.sprintf {|"%s" in "%s"|} contained container in
+    test_boolean (XString.has_substr container) (name, contained, expected)
+  in
+  ("has_substr", List.map test spec)
+
+let test_startswith =
+  let spec =
+    [
+      ("", "", true)
+    ; ("", "foo bar", true)
+    ; ("foofo", "foof", false)
+    ; ("foof", "foof", true)
+    ; ("f", "foof", true)
+    ; ("fo", "foof", true)
+    ; ("of", "foof", false)
+    ; ("ff", "foof", false)
+    ]
+  in
+  let test (contained, container, expected) =
+    let name = Printf.sprintf {|"%s" starts with "%s"|} container contained in
+    test_boolean (XString.startswith contained) (name, container, expected)
+  in
+  ("startswith", List.map test spec)
+
+let test_endswith =
+  let spec =
+    [
+      ("", "", true)
+    ; ("", "foo bar", true)
+    ; ("ofoof", "foof", false)
+    ; ("foof", "foof", true)
+    ; ("f", "foof", true)
+    ; ("fo", "foof", false)
+    ; ("of", "foof", true)
+    ; ("ff", "foof", false)
+    ]
+  in
+  let test (contained, container, expected) =
+    let name = Printf.sprintf {|"%s" ends in "%s"|} contained container in
+    test_boolean (XString.endswith contained) (name, container, expected)
+  in
+  ("endswith", List.map test spec)
+
+let test_rtrim =
+  let spec =
+    [
+      ("empty string", "", "")
+    ; ("single newline", "\n", "")
+    ; ("two newlines", "\n\n", "\n")
+    ; ("newline not ending", "\n ", "\n ")
```

Comment:
I would have expected some other cases, along the lines of

```
foo\n -> foo
```

(as an aside, where is this function actually useful? :confused:)

Reply:
It's used in several places, especially for reading from files